blackboxopt/optimizers/staged/bohb.py

```diff
@@ -279,20 +295,40 @@ def sample_configuration(self) -> Tuple[dict, dict]:
             kde_good = self.kde_models[fidelity]["good"]
             kde_bad = self.kde_models[fidelity]["bad"]
 
-            for _ in range(self.num_samples):
-                idx = np.random.randint(0, len(kde_good.data))
+            for i in range(self.num_samples):
+                idx = self._rng.integers(0, len(kde_good.data))
                 datum = kde_good.data[idx]
                 vector = sample_around_values(
                     datum,
                     kde_good.bw,
                     self.vartypes,
                     self.min_bandwidth,
                     self.bw_factor,
+                    rng=self._rng,
                 )
                 if vector is None:
                     continue
+
+                # Statsmodels KDE estimators relies on seeding through numpy's global
+                # state. We do this close to the evaluation of the PDF (`good`, `bad`)
+                # to increase robustness for multi threading.
+                # As we seed in a loop, we need to change it each iteration to not get
+                # the same random numbers each time.
+                # We also reset the np.random's global state, in case the user relies
+                # on it in other parts of the code and to not hide other determinism
+                # issues.
+                # TODO: Check github issue if there was progress and the seeding can be
+                # removed: https://github.com/statsmodels/statsmodels/issues/306
+                cached_rng_state = None
+                if self.seed:
+                    cached_rng_state = np.random.get_state()
+                    np.random.seed(self.seed + 1)
```

Comment:
Doesn't this line seed the rng for the statsmodel part always in the same way? I was expecting this seed to be changed somewhere to not get the same random numbers in each iteration. Hope I am missing something :)

Reply:
Ah, thanks! It has to be `self.seed + i` instead of `self.seed + 1`. Changed that temporarily for debugging and forgot to revert...